File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-bond-option-5-4/BondOptionStrike.cpp
// BondOptionStrike.cpp 
#include "BondOptionStrike.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

BondOptionStrike::BondOptionStrike(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //referenceSwapCurveNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* referenceSwapCurveNode = xmlNode->FirstChildElement("referenceSwapCurve");

   if(referenceSwapCurveNode){referenceSwapCurveIsNull_ = false;}
   else{referenceSwapCurveIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- referenceSwapCurveNode , address : " << referenceSwapCurveNode << std::endl;
   #endif
   if(referenceSwapCurveNode)
   {
      if(referenceSwapCurveNode->Attribute("href") || referenceSwapCurveNode->Attribute("id"))
      {
          if(referenceSwapCurveNode->Attribute("id"))
          {
             referenceSwapCurveIDRef_ = referenceSwapCurveNode->Attribute("id");
             referenceSwapCurve_ = boost::shared_ptr<ReferenceSwapCurve>(new ReferenceSwapCurve(referenceSwapCurveNode));
             referenceSwapCurve_->setID(referenceSwapCurveIDRef_);
             IDManager::instance().setID(referenceSwapCurveIDRef_,referenceSwapCurve_);
          }
          else if(referenceSwapCurveNode->Attribute("href")) { referenceSwapCurveIDRef_ = referenceSwapCurveNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { referenceSwapCurve_ = boost::shared_ptr<ReferenceSwapCurve>(new ReferenceSwapCurve(referenceSwapCurveNode));}
   }

   //priceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* priceNode = xmlNode->FirstChildElement("price");

   if(priceNode){priceIsNull_ = false;}
   else{priceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- priceNode , address : " << priceNode << std::endl;
   #endif
   if(priceNode)
   {
      if(priceNode->Attribute("href") || priceNode->Attribute("id"))
      {
          if(priceNode->Attribute("id"))
          {
             priceIDRef_ = priceNode->Attribute("id");
             price_ = boost::shared_ptr<OptionStrike>(new OptionStrike(priceNode));
             price_->setID(priceIDRef_);
             IDManager::instance().setID(priceIDRef_,price_);
          }
          else if(priceNode->Attribute("href")) { priceIDRef_ = priceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { price_ = boost::shared_ptr<OptionStrike>(new OptionStrike(priceNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<ReferenceSwapCurve> BondOptionStrike::getReferenceSwapCurve()
{
   if(!this->referenceSwapCurveIsNull_){
        if(referenceSwapCurveIDRef_ != ""){
             return boost::shared_static_cast<ReferenceSwapCurve>(IDManager::instance().getID(referenceSwapCurveIDRef_));
        }else{
             return this->referenceSwapCurve_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<ReferenceSwapCurve>();
   }
}
boost::shared_ptr<OptionStrike> BondOptionStrike::getPrice()
{
   if(!this->priceIsNull_){
        if(priceIDRef_ != ""){
             return boost::shared_static_cast<OptionStrike>(IDManager::instance().getID(priceIDRef_));
        }else{
             return this->price_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<OptionStrike>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-bond-option-5-4/BondOptionStrike.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class BondOptionStrike : ISerialized
    {
        public BondOptionStrike(XmlNode xmlNode)
        {
            XmlNode referenceSwapCurveNode = xmlNode.SelectSingleNode("referenceSwapCurve");
            
            if (referenceSwapCurveNode != null)
            {
                if (referenceSwapCurveNode.Attributes["href"] != null || referenceSwapCurveNode.Attributes["id"] != null) 
                {
                    if (referenceSwapCurveNode.Attributes["id"] != null) 
                    {
                        referenceSwapCurveIDRef_ = referenceSwapCurveNode.Attributes["id"].Value;
                        ReferenceSwapCurve ob = new ReferenceSwapCurve(referenceSwapCurveNode);
                        IDManager.SetID(referenceSwapCurveIDRef_, ob);
                    }
                    else if (referenceSwapCurveNode.Attributes["href"] != null)
                    {
                        referenceSwapCurveIDRef_ = referenceSwapCurveNode.Attributes["href"].Value;
                    }
                    else
                    {
                        referenceSwapCurve_ = new ReferenceSwapCurve(referenceSwapCurveNode);
                    }
                }
                else
                {
                    referenceSwapCurve_ = new ReferenceSwapCurve(referenceSwapCurveNode);
                }
            }
            
        
            XmlNode priceNode = xmlNode.SelectSingleNode("price");
            
            if (priceNode != null)
            {
                if (priceNode.Attributes["href"] != null || priceNode.Attributes["id"] != null) 
                {
                    if (priceNode.Attributes["id"] != null) 
                    {
                        priceIDRef_ = priceNode.Attributes["id"].Value;
                        OptionStrike ob = new OptionStrike(priceNode);
                        IDManager.SetID(priceIDRef_, ob);
                    }
                    else if (priceNode.Attributes["href"] != null)
                    {
                        priceIDRef_ = priceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        price_ = new OptionStrike(priceNode);
                    }
                }
                else
                {
                    price_ = new OptionStrike(priceNode);
                }
            }
            
        
        }
        
    
        #region ReferenceSwapCurve_
        private ReferenceSwapCurve referenceSwapCurve_;
        public ReferenceSwapCurve ReferenceSwapCurve_
        {
            get
            {
                if (this.referenceSwapCurve_ != null)
                {
                    return this.referenceSwapCurve_; 
                }
                else if (this.referenceSwapCurveIDRef_ != null)
                {
                    referenceSwapCurve_ = IDManager.getID(referenceSwapCurveIDRef_) as ReferenceSwapCurve;
                    return this.referenceSwapCurve_; 
                }
                else
                {
                      return this.referenceSwapCurve_; 
                }
            }
            set
            {
                if (this.referenceSwapCurve_ != value)
                {
                    this.referenceSwapCurve_ = value;
                }
            }
        }
        #endregion
        
        public string referenceSwapCurveIDRef_ { get; set; }
        #region Price_
        private OptionStrike price_;
        public OptionStrike Price_
        {
            get
            {
                if (this.price_ != null)
                {
                    return this.price_; 
                }
                else if (this.priceIDRef_ != null)
                {
                    price_ = IDManager.getID(priceIDRef_) as OptionStrike;
                    return this.price_; 
                }
                else
                {
                      return this.price_; 
                }
            }
            set
            {
                if (this.price_ != value)
                {
                    this.price_ = value;
                }
            }
        }
        #endregion
        
        public string priceIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

